### PR TITLE
More proto changes

### PR DESCRIPTION
**Pull request by AudriusButkevicius**

### CODE REVIEW

File: lib/protocol/message.go [hunk 1]
```diff
@@ -126,14 +127,24 @@ func (o *ClusterConfigMessage) GetOption(key string) string {
...
+	Devices         []Device // max:1000000
+	Flags           uint32
+	Options         []Option // max:64
 }
 
 type Device struct {
-	ID              []byte // max:32
+	ID              []byte   // max:32
+	Name            string   // max:64
+	Addresses       []string // max:64
```
Comment (calmh) [inline lib/protocol/message.go]: This needs a struct type, or we can't limit the length of the strings. (Or you could improve the xdr generator to handle this ;)

**Comment by calmh** [comment]
What's the overall goal here? I get the addresses-from-introducer part, but the rest?

**Comment by AudriusButkevicius** [comment]
DeviceName into its own field, but sending other stuff, such as wether the folder is read only or not can help later, for example deciding why something is out of sync.

**Comment by calmh** [comment]
OK

**Comment by AudriusButkevicius** [comment]
Fixed the tests

File: lib/model/model.go [hunk 2]
```diff
@@ -1484,31 +1491,43 @@ func (m *Model) numHashers(folder string) int {
...
 	m.fmut.RLock()
 	for _, folder := range m.deviceFolders[device] {
+		folderCfg := m.cfg.Folders()[folder]
 		cr := protocol.Folder{
-			ID: folder,
+			ID:              folder,
+			ReadOnly:        folderCfg.ReadOnly,
+			RescanIntervalS: folderCfg.RescanIntervalS,
+			IgnorePerms:     folderCfg.IgnorePerms,
+			AutoNormalize:   folderCfg.AutoNormalize,
```
Comment (calmh) [inline lib/model/model.go]: I don't think we gain much by exposing AutoNormalize; it's more of an internal fixup thing in my mind than something that should be exposed and documented at the protocol level...

**Comment by AudriusButkevicius** [comment]
Removed.

**Comment by AudriusButkevicius** [comment]
Now really fixed.

**Comment by calmh** [comment]
There's the above thing about `[]string` though

**Comment by AudriusButkevicius** [comment]
Ah yes

**Comment by AudriusButkevicius** [comment]
Updated.

File: lib/protocol/message_xdr.go [hunk 3]
```diff
@@ -700,6 +712,20 @@ Folder Structure:
 \                     ID (variable length)                      \
 /                                                               /
 +-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+
+|                    Read Only (V=0 or 1)                     |V|
```
Comment (calmh) [inline lib/protocol/message_xdr.go]: This is rather ugly from a wire protocol point of view. Can we not just define these in Flags?

File: lib/protocol/message.go [hunk 4]
```diff
@@ -126,14 +127,23 @@ func (o *ClusterConfigMessage) GetOption(key string) string {
 }
 
 type Folder struct {
-	ID      string   // max:256
-	Devices []Device // max:1000000
-	Flags   uint32
-	Options []Option // max:64
+	ID              string // max:256
+	ReadOnly        bool
+	RescanIntervalS int
```
Comment (calmh) [inline lib/protocol/message.go]: We should define this explicitly as int32 or int64 for the XDR encoding to be unambiguous.

**Comment by calmh** [comment]
Discovered further complaints on a closer look, sorry :dancer:

**Comment by AudriusButkevicius** [comment]
Updated

File: lib/model/model.go [hunk 5]
```diff
@@ -1465,31 +1472,50 @@ func (m *Model) numHashers(folder string) int {
...
 	}
 
 	m.fmut.RLock()
 	for _, folder := range m.deviceFolders[device] {
+		folderCfg := m.cfg.Folders()[folder]
 		cr := protocol.Folder{
-			ID: folder,
+			ID:              folder,
+			RescanIntervalS: uint32(folderCfg.RescanIntervalS),
+			Order:           uint32(folderCfg.Order),
```
Comment (calmh) [inline lib/model/model.go]: We're locking the various Order constants into the protocol here, so they should probably live in the protocol package from this point on?

**Comment by calmh** [comment]
I'm slowly running out of nits to pick here

**Comment by AudriusButkevicius** [comment]
So XDR does not support typedefs, and considers PullOrder a struct...

**Comment by calmh** [comment]
Sorry, I'm going to continue being a pain a little while longer here as this is essentially the protocol spec, which is (supposed to) be bigger than just our implementation.

Currently the rescan interval and pull order are quite specific to our implementation. What if inotify is in use or the implementation has no concept of a rescan (for example it's an app that uses beep to sync changes done by the user in real time)? Pull order might be something entirely different, like by internal ID or whatever. As we don't currently use these values I'm not sure we should force people to potentially make up lies to satisfy the protocol, now and forever into the future?

Also not convinced we need certname, as its in the the certificate that we've apparently already accepted to get to the point where we do this handshake at all. :)

Compression is defined in the protocol and could use just two flag bits: CompressionEnabled (at all or not) and CompressData (only metadata if enabled is set but this is unset, otherwise equivalent to compress always).

**Comment by AudriusButkevicius** [comment]
Certname can be for other devices that we are introducing.
Ill rip out others, given you are against it.

**Comment by AudriusButkevicius** [comment]
Not going to touch compression as we send it this way already.

**Pull request merged**